hibernate-core/src/main/java/org/hibernate/query/sql/internal/NativeQueryImpl.java

```diff
@@ -330,24 +330,32 @@ private void handleExplicitResultSetMapping() {
 				setTupleTransformerForResultType( resultType );
 			}
 			else {
-				checkResultType( resultType );
+				checkResultType( resultType, resultSetMapping );
 			}
 		}
 	}
 
-	private void checkResultType(Class<R> resultType) {
-		switch ( resultSetMapping.getNumberOfResultBuilders() ) {
-			case 0:
-				throw new IllegalArgumentException( "Named query exists, but did not specify a resultClass" );
-			case 1:
-				final Class<?> actualResultJavaType =
-						resultSetMapping.getResultBuilders().get( 0 ).getJavaType();
-				if ( actualResultJavaType != null && !resultType.isAssignableFrom( actualResultJavaType ) ) {
-					throw buildIncompatibleException( resultType, actualResultJavaType );
-				}
-				break;
-			default:
-				throw new IllegalArgumentException( "Cannot create TypedQuery for query with more than one return" );
+	private void checkResultType(Class<R> resultType, ResultSetMapping resultSetMapping) {
+		// resultType can be null if any of the deprecated methods were used to create the query
+		if ( resultType != null && !isResultTypeAlwaysAllowed( resultType )) {
+			switch ( resultSetMapping.getNumberOfResultBuilders() ) {
+				case 0:
+					if ( !resultSetMapping.isDynamic() ) {
+						throw new IllegalArgumentException( "Named query exists, but did not specify a resultClass" );
+					}
+					break;
+				case 1:
+					final Class<?> actualResultJavaType =
+							resultSetMapping.getResultBuilders().get( 0 ).getJavaType();
+					if ( actualResultJavaType != null && !resultType.isAssignableFrom( actualResultJavaType ) ) {
+						throw buildIncompatibleException( resultType, actualResultJavaType );
+					}
+					break;
+				default:
+						if ( resultType != Object[].class ) {
+							throw new IllegalArgumentException( "The declared return type for a multi-valued result set mapping should be Object[]" );
+						}
```

Comment:
If I understand correctly, I think this is "better":

```suggestion
						throw new IllegalArgumentException( "The declared return type for a multi-valued result set mapping should be Object[], Map, List, or Tuple" );
```

But I'm trying to understand what happens with implicit packaging as a record type.

I would have expected tests to fail with this implementation.

Reply:
> I would have expected tests to fail with this implementation.

Oh, well, they are failing. And the culprit is my `ImplicitInstantiationTest`. That's the one I expected to fail, so that's good.

Reply:
The rule is that it has to be a class with an appropriate constructor. So need to look for a constructor whose parameters match the types of the `ResultBuilder`s. If none exists, error.